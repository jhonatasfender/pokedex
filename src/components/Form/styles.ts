import Select from "react-select";
import styled from "styled-components";

export const Wrapper = styled.form`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  gap: 1rem;
  padding: 1rem;
  z-index: 4;
`;

export const SelectType = styled(Select)`
  max-width: 30rem;
  min-width: 30rem;
`;

export const Search = styled.input.attrs({ type: "text" })`
  border: 0.1rem solid hsl(0, 0%, 80%);
  border-radius: 0.4rem;
  height: 3.6rem;
  padding: 0 1rem;

  &:focus {
    outline: none;
  }
`;

export const WrapperCheckbox = styled.div`
  display: flex;
  align-items: center;
  gap: 1rem;
`;
