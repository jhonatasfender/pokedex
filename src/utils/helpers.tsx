import React from "react";
import { ThemeProvider } from "styled-components";

import { render, RenderResult } from "@testing-library/react";

import theme from "../style-global/theme";

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
