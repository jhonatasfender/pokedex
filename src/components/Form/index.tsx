import { useEffect } from "react";
import { Controller, FormProvider, useForm } from "react-hook-form";
import Select from "react-select";
import * as S from "./styles";

export type FormInputsProps = {
  search: string;
  type: string[];
  onlyLike: boolean;
  order: string;
};

type FormProps = {
  search: (form: Partial<FormInputsProps>) => void;
  listTypes: string[];
};

export default function Form({ search, listTypes }: FormProps) {
  const methods = useForm<FormInputsProps>();

  useEffect(() => {
    const subscription = methods.watch((val) => search(val as FormInputsProps));
    return () => subscription.unsubscribe();
  }, [methods, search]);

  return (
    <FormProvider {...methods}>
      <S.Wrapper>
        <S.Search type="text" {...methods.register("search")} />

        <Controller
          control={methods.control}
          {...methods.register("type")}
          render={({ field: { onChange, ref } }) => (
            <S.SelectType
              ref={ref}
              options={listTypes.map((type) => ({ value: type, label: type }))}
              onChange={(val) =>
                onChange((val as { value: string }[]).map((c) => c.value))
              }
              isMulti
            />
          )}
        />

        <Controller
          control={methods.control}
          {...methods.register("order")}
          render={({ field: { onChange, ref } }) => (
            <Select
              ref={ref}
              options={[
                { value: "name,asc", label: "Nome ASC" },
                { value: "name,desc", label: "Nome DESC" },
                {
                  value: "national_number,asc",
                  label: "Registro Nacional ASC",
                },
                {
                  value: "national_number,desc",
                  label: "Registro Nacional DESC",
                },
              ]}
              onChange={(val) => val && onChange(val.value)}
            />
          )}
        />

        <S.WrapperCheckbox>
          <label htmlFor="only-like">Filtrar somente os favoritados</label>

          <input
            type="checkbox"
            id="only-like"
            {...methods.register("onlyLike")}
          />
        </S.WrapperCheckbox>
      </S.Wrapper>
    </FormProvider>
  );
}
