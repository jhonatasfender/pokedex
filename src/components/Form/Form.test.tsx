import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { renderWithTheme } from "../../utils/helpers";
import Form from "./";

const Select = (props: any) => {
  const handleOnChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    if (!props.isMulti) {
      props.onChange({ value: e.target.value });
    } else {
      props.onChange([{ value: e.target.value }]);
    }
  };

  return (
    <select {...props} multiple={!!props.isMulti} onChange={handleOnChange}>
      {props.options.map((p: any) => (
        <option value={p.value} key={p.value}>
          {p.label}
        </option>
      ))}
    </select>
  );
};

jest.mock("react-select", () => ({
  __esModule: true,
  default: Select,
}));

describe("Form", () => {
  it("should initialize with default parameters", () => {
    const search = jest.fn();

    renderWithTheme(<Form search={search} listTypes={["test"]} />);

    const form = screen.getByRole("form");

    expect(form).toMatchSnapshot();
  });

  it("should call the Search function as soon as you click on check", async () => {
    const search = jest.fn();

    renderWithTheme(<Form search={search} listTypes={["test"]} />);

    userEvent.click(screen.getByRole("checkbox"));

    await waitFor(() => {
      expect(search).toHaveBeenCalledWith({
        onlyLike: true,
        order: undefined,
        search: "",
        type: undefined,
      });
    });
  });

  it("should call the Search function as soon as you select the sorting select option", async () => {
    const search = jest.fn();

    renderWithTheme(<Form search={search} listTypes={["test"]} />);

    const [select] = screen.getAllByRole("combobox");

    userEvent.selectOptions(select, ["national_number,desc"]);

    await waitFor(() => {
      expect(search).toHaveBeenCalledWith({
        onlyLike: false,
        order: "national_number,desc",
        search: "",
        type: undefined,
      });
    });
  });

  it("should call the Search function as soon as you select the type of select type", async () => {
    const search = jest.fn();

    renderWithTheme(<Form search={search} listTypes={["test"]} />);

    const [select] = screen.getAllByRole("listbox");

    userEvent.selectOptions(select, ["test"]);

    await waitFor(() => {
      expect(search).toHaveBeenCalledWith({
        onlyLike: false,
        order: undefined,
        search: "",
        type: ["test"],
      });
    });
  });
});
