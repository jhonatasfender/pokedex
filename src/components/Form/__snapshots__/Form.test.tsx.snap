// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should initialize with default parameters 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 1rem;
  padding: 1rem;
  z-index: 4;
}

.c2 {
  max-width: 30rem;
  min-width: 30rem;
}

.c1 {
  border: 0.1rem solid hsl(0,0%,80%);
  border-radius: 0.4rem;
  height: 3.6rem;
  padding: 0 1rem;
}

.c1:focus {
  outline: none;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1rem;
}

<form
  class="c0"
  role="form"
>
  <input
    class="c1"
    name="search"
    type="text"
  />
  <select
    class="c2"
    multiple=""
    options="[object Object]"
  >
    <option
      value="test"
    >
      test
    </option>
  </select>
  <select
    options="[object Object],[object Object],[object Object],[object Object]"
  >
    <option
      value="name,asc"
    >
      Nome ASC
    </option>
    <option
      value="name,desc"
    >
      Nome DESC
    </option>
    <option
      value="national_number,asc"
    >
      Registro Nacional ASC
    </option>
    <option
      value="national_number,desc"
    >
      Registro Nacional DESC
    </option>
  </select>
  <div
    class="c3"
  >
    <label
      for="only-like"
    >
      Filtrar somente os favoritados
    </label>
    <input
      id="only-like"
      name="onlyLike"
      type="checkbox"
    />
  </div>
</form>
`;
